tests/providers/amazon/aws/hooks/test_batch_waiters.py

```diff
@@ -0,0 +1,486 @@
+            "command": ["sleep", "10"],
+        },
+    )
+    assert resp["jobDefinitionName"] == job_definition_name
+    assert resp["jobDefinitionArn"]
+    job_definition_arn = resp["jobDefinitionArn"]
+    assert resp["revision"]
+    assert resp["jobDefinitionArn"].endswith(
+        "{0}:{1}".format(resp["jobDefinitionName"], resp["revision"])
+    )
+
+    infrastructure.vpc_id = vpc_id
+    infrastructure.subnet_id = subnet_id
+    infrastructure.security_group_id = sg_id
+    infrastructure.iam_arn = iam_arn
+    infrastructure.compute_env_name = compute_env_name
+    infrastructure.compute_env_arn = compute_env_arn
+    infrastructure.job_queue_name = job_queue_name
+    infrastructure.job_queue_arn = job_queue_arn
+    infrastructure.job_definition_name = job_definition_name
+    infrastructure.job_definition_arn = job_definition_arn
+    return infrastructure
+
+
+#
+# pytest tests
+#
+
+
+def test_aws_batch_waiters(aws_region):
+    assert inspect.isclass(AwsBatchWaiters)
+    batch_waiters = AwsBatchWaiters(region_name=aws_region)
+    assert isinstance(batch_waiters, AwsBatchWaiters)
+
+
+@mock_batch
+@mock_ec2
+@mock_ecs
+@mock_iam
+@mock_logs
+def test_aws_batch_job_waiting(aws_clients, aws_region, job_queue_name, job_definition_name):
+    """
+    Submit batch jobs and wait for various job status indicators or errors.
+    These batch job waiter tests can be slow and might need to be marked
+    for conditional skips if they take too long, although it seems to
+    run in about 30 sec to a minute.
+
+    .. note::
+        These tests have no control over how moto transitions the batch job status.
+
+    .. seealso::
+        - https://github.com/boto/botocore/blob/develop/botocore/waiter.py
+        - https://github.com/spulec/moto/blob/master/moto/batch/models.py#L360
+        - https://github.com/spulec/moto/blob/master/tests/test_batch/test_batch.py
+    """
+
+    aws_resources = batch_infrastructure(
+        aws_clients, aws_region, job_queue_name, job_definition_name
+    )
+    batch_waiters = AwsBatchWaiters(region_name=aws_resources.aws_region)
+
+    job_exists_waiter = batch_waiters.get_waiter("JobExists")
+    assert job_exists_waiter
+    assert isinstance(job_exists_waiter, botocore.waiter.Waiter)
+    assert job_exists_waiter.__class__.__name__ == "Batch.Waiter.JobExists"
+
+    job_running_waiter = batch_waiters.get_waiter("JobRunning")
+    assert job_running_waiter
+    assert isinstance(job_running_waiter, botocore.waiter.Waiter)
+    assert job_running_waiter.__class__.__name__ == "Batch.Waiter.JobRunning"
+
+    job_complete_waiter = batch_waiters.get_waiter("JobComplete")
+    assert job_complete_waiter
+    assert isinstance(job_complete_waiter, botocore.waiter.Waiter)
+    assert job_complete_waiter.__class__.__name__ == "Batch.Waiter.JobComplete"
+
+    # test waiting on a jobId that does not exist (this throws immediately)
+    with pytest.raises(botocore.exceptions.WaiterError) as err:
```

Comment:
I think we should use the more generic approach of checking for a raised exception via [assertRaises](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises), don't you think?

Reply:
Or is there a good reason to stick to pytests method?

Reply:
In the context of a pytest function, this is the only way that works, AFAIK.  The tests on the waiter that use moto mock infrastructure are based on work that already paved this way of testing code that requires AWS batch infrastructure (i.e. it's already approved in moto, for example).  AFAIK, this is just how it's done with pytest and as Airflow migrates to pytest, this pattern is likely the way it will be done.  (As an aside, I'd prefer that this PR is not held up as a proving ground for the best practices for Airflow pytest vs. unittest, but those debates should take place and the tests in this PR can be revised when those practices are worked out in total.)

Reply:
I'll take a liberty to resolve this and the related discussion below (there are dev-list discussions about these things that might result in changes that are out of scope here).